File: Codinator/Root/Playground/Generate Playground/NewPlaygroundViewController.m
//
//  NewPlaygroundViewController.m
//  Codinator
//
//  Created by Vladimir on 06/06/15.
//  Copyright (c) 2015 Vladimir Danila. All rights reserved.
//

#import "NewPlaygroundViewController.h"
#import "AppDelegate.h"
#import "Polaris.h"
#import "PlaygroundDocument.h"

@interface NewPlaygroundViewController (){
    UIColor *violett;
    __weak IBOutlet UIButton *nextButton;
}


@property (weak, nonatomic) IBOutlet UITextField *fileName;
@property (weak, nonatomic) IBOutlet UIButton *closeButton;


@end

@implementation NewPlaygroundViewController

- (void)viewDidLoad {
    [super viewDidLoad];



    violett = nextButton.backgroundColor;
    nextButton.backgroundColor = [UIColor grayColor];
    nextButton.enabled = NO;

    
    self.fileName.attributedPlaceholder = [[NSAttributedString alloc] initWithString:@"Playground name" attributes:@{NSForegroundColorAttributeName: [UIColor grayColor]}];


    self.fileName.layer.masksToBounds = true;
    self.fileName.layer.borderColor = [violett CGColor];
    self.fileName.layer.borderWidth = 1.0f;
    self.fileName.layer.cornerRadius = 5.0f;

    self.closeButton.layer.masksToBounds = YES;
    self.closeButton.layer.cornerRadius = 5;
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



#pragma mark - delegate


- (BOOL)textFieldShouldReturn:(UITextField *)textField{
    [textField resignFirstResponder];
    return true;
}

- (void)textFieldDidBeginEditing:(UITextField *)textField{
 
    self.fileName.layer.borderWidth = .0f;

    
}

- (void)textFieldDidEndEditing:(UITextField *)textField{
    
    if (textField.text.length == 0) {
        self.fileName.layer.masksToBounds = true;
        self.fileName.layer.borderColor = [violett CGColor];
        self.fileName.layer.borderWidth = 1.0f;
        self.fileName.layer.cornerRadius = 5.0f;
        
        nextButton.backgroundColor = [UIColor grayColor];
        nextButton.enabled = NO;

    }
    else{
        
        nextButton.enabled = YES;
        nextButton.backgroundColor = violett;
        
    }
    
}


#pragma mark - Generate


- (IBAction)playDidPush:(id)sender {
    
    //Root Path
    NSString *root = [AppDelegate storagePath];
    
    //Paths
    NSString *playgroudPaths = [root stringByAppendingPathComponent:@"Playground"];
    NSURL *fileUrl = [NSURL fileURLWithPath:[playgroudPaths stringByAppendingPathComponent:[NSString stringWithFormat:@"%@.cnPlay", self.fileName.text]]];
    
    PlaygroundDocument *document = [[PlaygroundDocument alloc] initWithFileURL:fileUrl];
    [document saveToURL:fileUrl forSaveOperation:UIDocumentSaveForCreating completionHandler:nil];
    
    
    
    NSString *html = [NSString stringWithFormat:
@"START \n\
    HEAD() \n\
        TITLE(\"%@\")TITLE \n\
        VIEWPORT(content: \"width=device-width\", initialScale: 1) \n\
        DESCRIPTION(\"A simple webpage written in Neuron\")         \n\
        AUTHOR(\"YOUR NAME\")    \n\
        IMPORT(CSS)   \n\
        IMPORT(JS)    \n\
    ()HEAD \n\
    BODY() \n\
    \n\
    H1(\"%@\")H1 \n\
    P(\"Hello world\")P   \n\
    \n\
    ()BODY \n\
END" ,self.fileName.text,self.fileName.text];
    
    [document.contents addObject:html];
    
    
    
    NSString *css = [NSString stringWithFormat:
                              @"/* Normalize.css brings consistency to browsers. \n\
                                 https://github.com/necolas/normalize.css */ \n\
                              \n\
                              @import url(http://cdn.jsdelivr.net/normalize/2.1.3/normalize.min.css); \n\
                              \n\
                              /* A fresh start */"];
 
    
    [document.contents addObject:css];
    

    
    NSString *js = [NSString stringWithFormat:
@"//JS file \n\
                        "];
    
    [document.contents addObject:js];


    
    
    [document saveToURL:fileUrl forSaveOperation:UIDocumentSaveForOverwriting completionHandler:^(BOOL success) {
        if (success) {
            [self dismissViewControllerAnimated:true completion:^{
                [[NSNotificationCenter defaultCenter] postNotificationName:@"createdProj" object:nil userInfo:nil];
                [[NSNotificationCenter defaultCenter] postNotificationName:@"reload" object:nil userInfo:nil];
                
            }];
        }
        else{
            UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:@"Failed to create Playground" preferredStyle:UIAlertControllerStyleAlert];
            UIAlertAction *closeAlert = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {
                
                [self dismissViewControllerAnimated:YES completion:nil];
                
            }];
            
            [alert addAction:closeAlert];
            [self presentViewController:alert animated:YES completion:nil];
            

            
        }
    }];
}


#pragma mark - extra buttons


- (IBAction)playgroundButtondPressed:(id)sender {
        
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Playground - noun:" message:@"A place where people can play and prototype stuff...." preferredStyle:UIAlertControllerStyleAlert];
                    UIAlertAction *closeAlert = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:nil];
    [alert addAction:closeAlert];
    [self presentViewController:alert animated:YES completion:nil];
    
    
}


- (IBAction)closeDidPush:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}


@end

File: Codinator/Root/Playground/Generate Playground/NewPlaygroundViewController.swift
//
//  NewPlaygroundViewController.swift
//  Codinator
//
//  Created by Vladimir Danila on 22/03/16.
//  Copyright © 2016 Vladimir Danila. All rights reserved.
//

import UIKit

class NewPlaygroundViewController: UIViewController, UITextFieldDelegate{

    @IBOutlet weak var fileNameTextField: UITextField!
    @IBOutlet weak var nextButton: UIButton!
    
    var violett: UIColor?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Save Violett color
        violett = nextButton.backgroundColor
        
        
        // Configure next Button
        nextButton.backgroundColor = UIColor.grayColor()
        nextButton.enabled = false

        // Configure fileNameTextField
        fileNameTextField.attributedPlaceholder = NSAttributedString(string: "Playground name", attributes: [NSForegroundColorAttributeName: UIColor.grayColor()])
        
        fileNameTextField.layer.masksToBounds = true
        fileNameTextField.layer.borderColor = violett?.CGColor
        fileNameTextField.layer.borderWidth = 1.0
        fileNameTextField.layer.cornerRadius = 5
        
        fileNameTextField.addTarget(self, action: #selector(NewPlaygroundViewController.textFieldDidChange), forControlEvents: UIControlEvents.EditingChanged)

        
        self.nextButton.backgroundColor = UIColor.grayColor()
        self.nextButton.setTitleColor(UIColor.darkGrayColor(), forState: UIControlState.Normal)
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    // MARK: - Buttons

    
    // Generate a new File
    @IBAction func nextDidPush(sender: AnyObject) {
        let document = PlaygroundFileCreator.generatePlaygroundFileWithName(fileNameTextField.text!)
        let url = PlaygroundFileCreator.fileUrlForPlaygroundWithName(fileNameTextField.text!)

        document.saveToURL(url, forSaveOperation: .ForOverwriting) { (success) in
            if success {
                self.dismissViewControllerAnimated(true, completion: { 
                    NSNotificationCenter.defaultCenter().postNotificationName("createdProj", object: nil, userInfo: nil)
                    NSNotificationCenter.defaultCenter().postNotificationName("reload", object: self, userInfo: nil)
                })
            }
            else {
                let alertController = UIAlertController(title: "Error", message: "Failed to create Playground", preferredStyle: .Alert)
                let cancelAction = UIAlertAction(title: "Ok", style: .Cancel, handler: { (UIAlertAction) in
                    self.dismissViewControllerAnimated(true, completion: nil)
                })
                alertController.addAction(cancelAction)
                
                self.presentViewController(alertController, animated: true, completion: nil)
            }
        }
        
    }
    
    
    // Close View 
    @IBAction func closeDidPush(sender: AnyObject) {
        self.dismissViewControllerAnimated(true, completion: nil)
    }

    
    
    // MARK: - Textfield Delegate
    
    func textFieldDidChange() {
        if fileNameTextField.text?.isEmpty == true {
            fileNameTextField.layer.masksToBounds = true
            fileNameTextField.layer.borderColor = violett?.CGColor
            fileNameTextField.layer.borderWidth = 1
            fileNameTextField.layer.cornerRadius = 5
            
            // Textfield is emtpy so disable NextButton
            nextButton.enabled = false
            
            UIView.animateWithDuration(0.2, animations: { 
                self.nextButton.backgroundColor = UIColor.grayColor()
                self.nextButton.setTitleColor(UIColor.darkGrayColor(), forState: UIControlState.Normal)

            })
        }
        else {
            
            // Textfield is not empty so enable NextButton
            nextButton.enabled = true
            
            UIView.animateWithDuration(0.2, animations: {
                self.nextButton.backgroundColor = self.violett
                self.nextButton.setTitleColor(self.view.tintColor, forState: UIControlState.Normal)
            })
        }
    }
    
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        textField.resignFirstResponder()
        return false
    }
    

}
